From: Andi Kleen <ak@linux.intel.com>
Subject: HWMON: Convert via-cputemp to x86 cpuid autoprobing
References: http://lists.opensuse.org/opensuse-kernel/2011-11/msg00075.html
Patch-Mainline: 267fc9788d0cdb77edafb506063f06961e1418f5 undefined


Signed-off-by: Thomas Renninger <trenn@suse.de>

Use the new x86 cpuid autoprobe interface.

Cc: Jean Delvare <khali@linux-fr.org>
Cc: Guenter Roeck <guenter.roeck@ericsson.com>
Signed-off-by: Andi Kleen <ak@linux.intel.com>
Signed-off-by: Thomas Renninger <trenn@suse.de>
Acked-by: H. Peter Anvin <hpa@zytor.com>
Signed-off-by: Greg Kroah-Hartman <gregkh@suse.de>

Automatically created from "patches.arch/004-x86_autoload.patch" by xen-port-patches.py

--- head.orig/drivers/hwmon/via-cputemp-xen.c	2011-09-12 11:45:20.000000000 +0200
+++ head/drivers/hwmon/via-cputemp-xen.c	2012-02-16 14:35:20.000000000 +0100
@@ -36,6 +36,7 @@
 #include <linux/platform_device.h>
 #include <linux/cpu.h>
 #include <asm/msr.h>
+#include <asm/cpu_device_id.h>
 #include <xen/pcpu.h>
 #include "../xen/core/domctl.h"
 
@@ -331,6 +332,14 @@ static struct notifier_block via_cputemp
 	.notifier_call = via_cputemp_cpu_callback,
 };
 
+static const struct x86_cpu_id cputemp_ids[] = {
+	{ X86_VENDOR_CENTAUR, 6, 0xa, }, /* C7 A */
+	{ X86_VENDOR_CENTAUR, 6, 0xd, }, /* C7 D */
+	{ X86_VENDOR_CENTAUR, 6, 0xf, }, /* Nano */
+	{}
+};
+MODULE_DEVICE_TABLE(x86cpu, cputemp_ids);
+
 static int __init via_cputemp_init(void)
 {
 	int err;
@@ -338,11 +347,8 @@ static int __init via_cputemp_init(void)
 	if (!is_initial_xendomain())
 		return -ENODEV;
 
-	if (cpu_data(0).x86_vendor != X86_VENDOR_CENTAUR) {
-		printk(KERN_DEBUG DRVNAME ": Not a VIA CPU\n");
-		err = -ENODEV;
-		goto exit;
-	}
+	if (!x86_match_cpu(cputemp_ids))
+		return -ENODEV;
 
 	err = platform_driver_register(&via_cputemp_driver);
 	if (err)
